### PR TITLE
Decouple from Buzz

### PR DESCRIPTION
**Pull request by fbourigault**
As discussed in #171 this replace the buzz client by [httplug](http://docs.php-http.org/en/latest/httplug/introduction.html).

I tried to implement this as it was done by @Nyholm for https://github.com/KnpLabs/php-github-api. This remove all the HttpClient stuff which was built on the top of Buzz. Instead it use Httplug standard plugins with a few custom ones for error handling and authentication.

I set PHP required version to 5.6+ as discussed in #171.

This **IS** a breaking change.

## To do
* [x] Update documentation.
* [x] Describe breaking changes.
* [x] Add tests.
* [x] Do intensive tests against real gitlab repositories.
* [x] Update ResultPager.
* [x] https://github.com/m4tthumphrey/php-gitlab-api/pull/191#issuecomment-308155979

### CODE REVIEW

File: lib/Gitlab/Api/AbstractApi.php [hunk 1]
```diff
@@ -1,6 +1,11 @@
 <?php namespace Gitlab\Api;
 
 use Gitlab\Client;
+use Gitlab\HttpClient\Message\ResponseMediator;
+use function GuzzleHttp\Psr7\stream_for;
```
Comment (Nyholm) [inline lib/Gitlab/Api/AbstractApi.php]: You need to remove this line. It will couple us to Guzzle. You should use the `StreamFactoryDiscovery` instead.

**Comment by fbourigault** [comment]
@Nyholm I fixed the StreamFactory related issue. About the factory function, as there is many authentication methods, I don't know which one I should promote over the others.

I also added tests on some new classes, improved the `Gitlab\Client` code to limit the amount of breaking changes. I re-introduced the user-agent with a getter and a setter instead of the options array.

ResultPager is now fixed, but still waiting for unit tests.

File: lib/Gitlab/Client.php [hunk 2]
```diff
@@ -54,48 +58,62 @@ class Client
...
 
     /**
-     * @var array
+     * @var string
      */
-    private $options = array(
-        'user_agent'  => 'php-gitlab-api (http://github.com/m4tthumphrey/php-gitlab-api)',
-        'timeout'     => 60
-    );
+    private $userAgent = 'php-gitlab-api (http://github.com/m4tthumphrey/php-gitlab-api)';
```
Comment (Nyholm) [inline lib/Gitlab/Client.php]: Do we really need this in a property? We never read from it (as I can see)

Comment (fbourigault) [inline lib/Gitlab/Client.php]: Users where able to read it through `getOption('user_agent')`. I added it to limit the breaking changes surface.

**Comment by fbourigault** [comment]
I removed useless methods kept to limit breaking changes.
I documented the breaking changes in `UPGRADE.md` and also updated `README.md`.

**Comment by fbourigault** [comment]
ResultPager is currently broken. Waiting for https://github.com/php-http/client-common/pull/74 and a release.

**Comment by fbourigault** [comment]
I no longer require https://github.com/php-http/client-common/pull/74. I ended with a ApiVersion plugin which add the `/api/v3/` part when required.

This PR is definitively ready for review.

**Comment by m1guelpf** [comment]
ping @m4tthumphrey

File: README.md [hunk 3]
```diff
@@ -7,25 +7,14 @@ Based on [php-github-api](https://github.com/m4tthumphrey/php-github-api) and co
...
-2. Add the following to your require block in composer.json config.
-
-    > Note: be careful when using the `dev-master` tag as this may have unexpected results depending on your version of
-    Gitlab. See the Versioning section below for more information.
-
-    `php composer.phar require m4tthumphrey/php-gitlab-api:dev-master`
-
-3. Include Composer's autoloader:
+```bash
+composer require m4tthumphrey/php-gitlab-api php-http/guzzle6-adapter
```
Comment (Nyholm) [inline README.md]: We should add php-http/message here. We could also add a link to http://docs.php-http.org/en/latest/httplug/users.html

File: lib/Gitlab/Client.php [hunk 4]
```diff
@@ -54,48 +60,50 @@ class Client
...
-        $this->baseUrl     = $baseUrl;
-        $this->httpClient  = new HttpClient($this->baseUrl, $this->options, $httpClient);
+    /**
+     * Create a Gitlab\Client using a HttpClient.
+     *
+     * @param HttpClient $httpClient
+     *
+     * @return Client
+     */
+    public static function createWithHttpClient(HttpClient $httpClient)
```
Comment (Nyholm) [inline lib/Gitlab/Client.php]: Hm, I know I asked before. Maybe create a factory function for URL?

`Client::create($url)`?

File: lib/Gitlab/Client.php [hunk 5]
```diff
@@ -250,50 +243,34 @@ public function clearHeaders()
...
-     * @param string $name
-     *
-     * @return mixed
-     *
-     * @throws InvalidArgumentException
+     * @param string $api
+     * @return AbstractApi
      */
-    public function getOption($name)
+    public function __get($api)
```
Comment (Nyholm) [inline lib/Gitlab/Client.php]: FYI, https://github.com/mailgun/mailgun-php/pull/192#discussion_r84417626

We decided to remove it in Mailgun api

File: lib/Gitlab/HttpClient/Plugin/ApiVersion.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+class ApiVersion implements Plugin
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
+    {
+        $uri = $request->getUri();
+
+        if (substr($uri->getPath(), 0, 8) !== '/api/v3/') {
```
Comment (Nyholm) [inline lib/Gitlab/HttpClient/Plugin/ApiVersion.php]: Hm, But what if there get new endpoints with `/api/v4/`?

Comment (fbourigault) [inline lib/Gitlab/HttpClient/Plugin/ApiVersion.php]: There are no plans to support multiple versions in a single release. V3 api will be discontinued in august, so the plan is to replace this by v4.

File: lib/Gitlab/Client.php (hunk 4 above)
Comment (fbourigault) [inline lib/Gitlab/Client.php]: Looks interesting as many companies/users run their own GitLab instance.

File: lib/Gitlab/Client.php (hunk 5 above)
Comment (fbourigault) [inline lib/Gitlab/Client.php]: Humm. This should not be in the diff. I will give it a look. Moreover, I also prefer writing methods and remove the magic method and huge comment in the top of the file. It's the same amount of work!

Comment (fbourigault) [inline lib/Gitlab/Client.php]: And this will be part of an other PR.

File: lib/Gitlab/Client.php (hunk 4 above)
Comment (Nyholm) [inline lib/Gitlab/Client.php]: Yeah, exactly.

```php
$client = new \Gitlab\Client();
$client->setUrl('http://git.yourdomain.com');

// Could be the same as
$client = Gitlab\Client::create('http://git.yourdomain.com');
```

**Comment by fbourigault** [comment]
What about removing `Client::setUserAgent` and documenting how to customize it like https://github.com/KnpLabs/php-github-api/blob/master/doc/customize.md#configure-the-http-client?

File: README.md (hunk 3 above)
Comment (fbourigault) [inline README.md]: When using `php-http/guzzle6-adapter`, there is no need for `php-http/message`, right?

**Comment by Nyholm** [comment]
👍

**Comment by fbourigault** [comment]
As we are breaking, I cleaned  a bit more the `Gitlab\Client` public API. I will update `UPGRADE.md` and add a `doc/customize.md` to document how to set the client timeout, how to customize the user-agent and how to add custom headers.

**Comment by fbourigault** [comment]
Documentation is now up-to-date.

**Comment by m1guelpf** [comment]
@fbourigault Is this ready now?

**Comment by fbourigault** [comment]
Yes! I will change things only if a reviewer can convince me to do so!

File: README.md (hunk 3 above)
Comment (Nyholm) [inline README.md]: ah, okey.

File: UPGRADE.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# UPGRADE FROM 8.0 to 9.0
+
+Since 9.0, lib no longer use buzz 0.7+, instead it has an HTTPlug abstraction layer.
+
+## `Gitlab\Client` changes
+
+* The constructor no longer allow to specify base url. Use `setUrl` instead.
```
Comment (Nyholm) [inline UPGRADE.md]: Use `setUrl` or `Client::create` instead.

File: lib/Gitlab/Client.php [hunk 8]
```diff
@@ -54,48 +60,63 @@ class Client
...
      */
-    private $httpClient;
+    public function __construct(Builder $httpClientBuilder = null)
+    {
+        $this->responseHistory = new History();
+        $this->httpClientBuilder = $httpClientBuilder ?: new Builder();
+
+        $this->httpClientBuilder->addPlugin(new GitlabExceptionThrower());
+        $this->httpClientBuilder->addPlugin(new HistoryPlugin($this->responseHistory));
+        $this->httpClientBuilder->addPlugin(new RedirectPlugin());
```
Comment (Nyholm) [inline lib/Gitlab/Client.php]: Do we really need this?

File: lib/Gitlab/HttpClient/Message/ResponseMediator.php [hunk 9]
```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Gitlab\HttpClient\Message;
+
+use Psr\Http\Message\ResponseInterface;
+
```
Comment (Nyholm) [inline lib/Gitlab/HttpClient/Message/ResponseMediator.php]: Maybe a class comment?

**Comment by fbourigault** [comment]
@m4tthumphrey, do you agree with these changes?

**Comment by fbourigault** [comment]
Is it ok if I move forward with next release and merge this PR by the end of this week? ping @m4tthumphrey @jubianchi @radutopala

**Comment by fbourigault** [comment]
https://twitter.com/m4tthumphrey/status/883245976198275072

**Comment by m1guelpf** [comment]
@fbourigault Did you tag a new release?

**Comment by fbourigault** [comment]
Not yet, we have to make the library v4 only.

**Comment by m1guelpf** [comment]
@fbourigault Do you have a task list/milestone?

**Comment by fbourigault** [comment]
I will post a task list today with some "instructions" to be efficient.

**Comment by m1guelpf** [comment]
@fbourigault Please ping me then :smile:

**Comment by m4tthumphrey** [comment]
Amazing @fbourigault! Top work!

**Pull request merged**